### ISSUE TITLE
Cannot save Lambda layer with custom masking

### ISSUE THREAD

**Issue by saulvargas** [keras-team/keras#12283]
The following code, which contains a Lambda layer with custom masking logic, works fine but cannot be saved:

```import numpy as np
from numpy.testing import assert_array_almost_equal

import keras.backend as K
from keras import Model
from keras.layers import Input, Embedding, Lambda


def _mask(_1, _2):
    return None


def _average(x, mask):
    w = K.cast(mask, "float32")
    w = w / (K.sum(w, axis=1, keepdims=True) + 1e-16)

    return K.sum(x * K.expand_dims(w, -1), axis=1, keepdims=True)


n_ids = 3
k = 2
vectors = np.array([[0.0, 0.0], [0.0, 1.0], [1.0, 0.0], [1.0, 1.0]])

x = Input(shape=(None,), name='ids', dtype='int32')
y = Embedding(input_dim=n_ids + 1, output_dim=k, mask_zero=True, weights=[vectors])(x)
z = Lambda(_average, mask=_mask, output_shape=(None, k))(y)

model = Model(inputs=x, outputs=z)

expected = np.array([1.0, 0.5])

assert_array_almost_equal(expected, model.predict([[[2, 3]]]).squeeze())
assert_array_almost_equal(expected, model.predict([[[0, 2, 3]]]).squeeze())

model.save("model.h5")
```

After some debugging, I have determined that the problem can be partially fixed by applying this commit: https://github.com/saulvargas/keras/commit/7bddc1a9c50df582ee87a77943d04c0f97b3ae17 Esentially it prevents tensorflow tensors to be part of the Lambda layer via the `arguments` object variable.

Unfortunately the function doing the custom masking is not saved so, when loading the saved, model, this would not work properly (unless the output mask is not used downstream).

**Comment by gabrieldemarmiesse** [comment]
You should not pass multiple tensors to lambda layers like that. See the docs for this example:

```python
# add a layer that returns the hadamard product
# and sum of it from two input tensors

def hadamard_product_sum(tensors):
    out1 = tensors[0] * tensors[1]
    out2 = K.sum(out1, axis=-1)
    return [out1, out2]

def hadamard_product_sum_output_shape(input_shapes):
    shape1 = list(input_shapes[0])
    shape2 = list(input_shapes[1])
    assert shape1 == shape2  # else hadamard product isn't possible
    return [tuple(shape1), tuple(shape2[:-1])]

x1 = Dense(32)(input_1)
x2 = Dense(32)(input_2)
layer = Lambda(hadamard_product_sum, hadamard_product_sum_output_shape)
x_hadamard, x_sum = layer([x1, x2])
```

**Comment by saulvargas** [comment]
Hi @gabrieldemarmiesse thank you for checking this issue. I am afraid I haven't been clear enough in my explanation as your answer is somewhat unrelated. I am aware of how to pass two input tensors to a layer.

In this issue, I am referring to the standard case of a Lambda layer that takes a lambda function that takes not only the "normal" input (a tensor or an array of them), BUT also a mask: https://github.com/keras-team/keras/blob/master/keras/layers/core.py#L708 Please do not worry about the silly example I posted, it is just to show the general issue.

**Comment by gabrieldemarmiesse** [comment]
My bad. I think the documentation is wrong. The mask parameter is supposed to be a boolean, and the mask is supposed to be specified only when you call the layer (like other layers supporting masking). There is indeed something fishy going on, but I don't believe that tensors should be arguments of layers contructors.

**Comment by gabrieldemarmiesse** [comment]
Flagging as bug because it needs either a fix, either a clarification of the docs.
